### PR TITLE
Add implementation of `std::ranges::partial_sort`

### PR DESCRIPTION
**Pull request by SergeyKopienko**
## Summary

This PR adds a parallel range-based implementation of `std::ranges::partial_sort` to oneDPL, following the same design already used for the other range-based sorting algorithms.

Reference: C++ standard, [partial.sort](https://eel.is/c++draft/alg.sort#partial.sort).

## Details

`oneapi::dpl::ranges::partial_sort` reorders the elements of a range so that the sub-range `[begin(r), middle)` contains the smallest `middle - begin(r)` elements of the whole range in sorted order (with respect to the projected comparator), while the order of the remaining elements `[middle, end(r))` is unspecified. It returns a `borrowed_iterator_t` equal to `end(r)`.

Declared signature (matches `documentation/specification/parallel_api/parallel_range_api.rst`):

```cpp
template <typename ExecutionPolicy, typename R, typename Comp = std::ranges::less, typename Proj = std::identity>
  requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<ExecutionPolicy>> &&
           std::ranges::sized_range<R> &&
           std::sortable<std::ranges::iterator_t<R>, Comp, Proj>
std::ranges::borrowed_iterator_t<R>
partial_sort(ExecutionPolicy&& pol, R&& r, std::ranges::iterator_t<R> middle, Comp comp = {}, Proj proj = {});
```

## Conformance with [partial.sort](https://eel.is/c++draft/alg.sort#partial.sort)

The implementation is consistent with the standard specification of `ranges::partial_sort`:

- **Effects**: the first `middle - begin(r)` elements in sorted order are placed into `[begin(r), middle)`; the order of `[middle, end(r))` is unspecified. The serial path forwards directly to `std::ranges::partial_sort`; the parallel and device paths reuse the existing `__pattern_partial_sort()` implementation with a `__binary_op<_Comp, _Proj, _Proj>` comparator.
- **Return value**: returns `end(r)` as a `borrowed_iterator_t` on every code path.
- **Requirements**: the standard requires `random_access_range` and `sortable<iterator_t<R>, Comp, Proj>`. oneDPL additionally constrains the input with `sized_range`, as required by the parallel/device backends; this is the established convention for all oneDPL parallel range algorithms and does not change the observable semantics.

### CODE REVIEW

File: test/parallel_api/ranges/std_ranges_partial_sort.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+enum class middle_pos
+{
+    first,
+    half,
+    last
+};
+
+template <middle_pos Pos, std::ranges::range R>
+std::ranges::borrowed_iterator_t<R>
+get_middle(R&& r)
```
Comment (dmitriy-sobolev) [inline test/parallel_api/ranges/std_ranges_partial_sort.pass.cpp]: middle's type is `std::ranges::iterator_t` rather than `std::ranges::borrowed_iterator_t`.
It probably changes nothing in practice though.

File: test/parallel_api/ranges/std_ranges_partial_sort.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    // Boundary case: middle == begin(r), the algorithm is a no-op
+    auto partial_sort_none_algo = partial_sort_fn<middle_pos::first>{};
+    auto partial_sort_none_checker = [](auto&& r, auto&&... args) -> std::ranges::borrowed_iterator_t<decltype(r)> {
+        auto middle = get_middle<middle_pos::first>(r);
+        return std::ranges::partial_sort(std::forward<decltype(r)>(r), middle, std::forward<decltype(args)>(args)...);
+    };
+
+    test_range_algo<8>{big_sz}(partial_sort_none_algo, partial_sort_none_checker);
+    test_range_algo<9>{}(partial_sort_none_algo, partial_sort_none_checker, std::ranges::greater{}, proj);
```
Comment (dmitriy-sobolev) [inline test/parallel_api/ranges/std_ranges_partial_sort.pass.cpp]: No reason to check this corner case with different comparators and projections. This one also does not need many elements. 

```suggestion
    test_range_algo<8>{}(partial_sort_none_algo, partial_sort_none_checker);
```

File: test/parallel_api/ranges/std_ranges_partial_sort.pass.cpp [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    auto partial_sort_algo = partial_sort_fn<middle_pos::half>{};
+
+    auto partial_sort_checker = [](auto&& r, auto&&... args) -> std::ranges::borrowed_iterator_t<decltype(r)> {
+        auto middle = get_middle<middle_pos::half>(r);
+        return std::ranges::partial_sort(std::forward<decltype(r)>(r), middle, std::forward<decltype(args)>(args)...);
+    };
+
+    test_range_algo<0>{big_sz}(partial_sort_algo, partial_sort_checker);
+    test_range_algo<1>{}(partial_sort_algo, partial_sort_checker, std::ranges::less{});
```
Comment (dmitriy-sobolev) [inline test/parallel_api/ranges/std_ranges_partial_sort.pass.cpp]: Let's test the large number of elements with greater{} to cover unsorted sequence instead of sorted.

```suggestion
    test_range_algo<0>{}(partial_sort_algo, partial_sort_checker);
    test_range_algo<1>{big_sz}(partial_sort_algo, partial_sort_checker, std::ranges::greater{});
```

File: include/oneapi/dpl/pstl/algorithm_ranges_impl.h [hunk 4]
```diff
@@ -465,6 +465,37 @@ __pattern_sort_ranges(__serial_tag</*IsVector*/ std::false_type>, _ExecutionPoli
...
+__pattern_partial_sort_ranges(_Tag __tag, _ExecutionPolicy&& __exec, _R&& __r, std::ranges::iterator_t<_R> __middle,
+                              _Comp __comp, _Proj __proj)
+{
+    static_assert(__is_parallel_tag_v<_Tag> || typename _Tag::__is_vector{});
+
+    auto [__first, __last] = oneapi::dpl::__ranges::__bounds(__r);
+
+    if (__middle == __first)
+        return __last;
+
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/algorithm_ranges_impl.h]:
```suggestion
```

__pattern_partial_sort already does this check:

```
    const auto __n = __middle - __first;
    if (__n == 0)
        return;
```

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h [hunk 5]
```diff
@@ -1504,6 +1504,25 @@ __pattern_sort_ranges(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec
...
+template <typename _BackendTag, typename _ExecutionPolicy, typename _R, typename _Comp, typename _Proj>
+std::ranges::borrowed_iterator_t<_R>
+__pattern_partial_sort_ranges(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _R&& __r,
+                              std::ranges::iterator_t<_R> __middle, _Comp __comp, _Proj __proj)
+{
+    auto [__first, __last] = oneapi::dpl::__ranges::__bounds(__r);
+
+    if (__middle == __first)
+        return __last;
+
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h]: No reason to do this short-circuit since you have already added it to `__pattern_partial_sort`

File: test/parallel_api/ranges/std_ranges_partial_sort.pass.cpp (hunk 1 above)
Comment (SergeyKopienko) [reply test/parallel_api/ranges/std_ranges_partial_sort.pass.cpp]: Fixed

File: include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h (hunk 5 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h]: Fixed

File: include/oneapi/dpl/pstl/algorithm_ranges_impl.h (hunk 4 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/algorithm_ranges_impl.h]: Fixed

**Review by dmitriy-sobolev — DISMISSED** [review]
Looks good given the CI passes.

**Review by dmitriy-sobolev — APPROVED** [review]
